### PR TITLE
Add Node.list/2, matching erlang:nodes/2

### PR DESCRIPTION
**Pull request by v0idpwn**
Per #11985

### CODE REVIEW

File: lib/elixir/lib/node.ex [hunk 1]
```diff
@@ -92,12 +93,49 @@ defmodule Node do
...
+    [{:connection_id, boolean()} | {:node_type, boolean()}]
+  ) :: [{t, [{:connection_id, integer() | nil} | {:node_type, state()}]}]
+  def list(args, info_opts) do
+    args
+    |> :erlang.nodes(Map.new(info_opts))
+    |> Enum.map(fn {node, info} ->
+      info = Enum.map(info, fn
+        {:connection_id, :undefined} -> {:connection_id, nil}
+        other -> other
+      end)
```
Comment (sabiwara) [inline lib/elixir/lib/node.ex:133]: Perhaps returning a map is more convenient here?

```suggestion
      info = case info do
         %{:connection_id: :undefined} -> %{info | connection_id: nil}
        other -> other
      end
```

File: lib/elixir/lib/node.ex [hunk 2]
```diff
@@ -92,12 +93,49 @@ defmodule Node do
...
   end
 
+  @doc """
+  Returns a list of nodes according to argument given, with extra info.
+
+  The second argument is used to request additional information with the
+  following options:
+
+  * `:connection_id` - boolean. Returns the connection id from each node.
+  * `:node_type` - boolean. Returns the node type (or state) of each node.
```
Comment (sabiwara) [inline lib/elixir/lib/node.ex:108]: Could we also include the defaults?

File: lib/elixir/lib/node.ex [hunk 3]
```diff
@@ -92,12 +93,49 @@ defmodule Node do
...
+
+  The additional information is returned as a keyword.
+
+  ## Examples
+
+      iex> Node.list([:connected, :this], [connection_id: true])
```
Comment (sabiwara) [inline lib/elixir/lib/node.ex:114]: It's probably better to use the syntax that the formatter would give?
```suggestion
      iex> Node.list([:connected, :this], connection_id: true)
```

File: lib/elixir/lib/node.ex [hunk 4]
```diff
@@ -92,12 +93,49 @@ defmodule Node do
...
+
+      iex> Node.list([:connected, :this], [connection_id: true, node_type: true])
+      [myself@localhost: [connection_id: nil, node_type: :this], node_a@localhost: [connection_id: 1, node_type: :visible]]
+
+  For more information, see `:erlang.nodes/2`.
+  """
+  @spec list(
+    state | [state],
+    [{:connection_id, boolean()} | {:node_type, boolean()}]
+  ) :: [{t, [{:connection_id, integer() | nil} | {:node_type, state()}]}]
```
Comment (josevalim) [inline lib/elixir/lib/node.ex:125]: Use keyword lists in the typespecs too please.

File: lib/elixir/lib/node.ex (hunk 1 above)
Comment (josevalim) [reply lib/elixir/lib/node.ex:133]: I am not the biggest fan we need to do an additional traversal here. I am wondering if we really need to provide a wrapper around this function? Perhaps we just document there is a function that returns more metadata in Erlang itself?

Comment (v0idpwn) [reply lib/elixir/lib/node.ex:133]: I was also questioning this, but I've seen it in #11985 and thought it was a low hanging fruit 😄, but now I'm looking at it again and it was removed (?)

Comment (josevalim) [reply lib/elixir/lib/node.ex:133]: I removed it yesterday but I am glad you sent the PR anyway, because it helped confirm the “feeling” :)

**Comment by josevalim** [comment]
Closing then as per discussion above!

**Pull request closed without merging**